### PR TITLE
Refactor unleash implementation

### PR DESCRIPTION
**Pull request by janboll**
Keep the client, do not create one per Request. This reduces the number
of requests to unleash. This reduces likelyhood of DNS errors due to
failed lookups during cluster upgrades.

Related ticket: APPSRE-5719

### CODE REVIEW

**Comment by janboll** [comment]
Tested by comparing results with old and new implementation using a couple of feature toggles. Including `openshift-resources-native-client`

File: reconcile/utils/unleash.py [hunk 1]
```diff
@@ -1,117 +1,94 @@
...
-from UnleashClient import UnleashClient
+from UnleashClient import UnleashClient, BaseCache
 from UnleashClient import strategies
+from UnleashClient.strategies import Strategy
 
-from reconcile.utils.defer import defer
 from reconcile.utils.helpers import toggle_logger
 
-log_lock = threading.Lock()
+client_local = threading.local()
```
Comment (geoberle) [inline reconcile/utils/unleash.py:11]: So I understand correctly that there will be one unleash client per thread?

Comment (janboll) [reply reconcile/utils/unleash.py:11]: Correct, I added a test to make that clearer.

File: reconcile/utils/oc.py [hunk 2]
```diff
@@ -1231,7 +1231,7 @@ def __new__(
             use_native = use_native.lower() in ["true", "yes"]
         else:
             enable_toggle = "openshift-resources-native-client"
-            strategies = get_feature_toggle_strategies(enable_toggle, ["perCluster"])
+            strategies = get_feature_toggle_strategies(enable_toggle)
```
Comment (maorfr) [inline reconcile/utils/oc.py:1234]: why is this being removed?

File: reconcile/utils/unleash.py (hunk 1 above)
Comment (geoberle) [reply reconcile/utils/unleash.py:11]: What is the reason to have one per thread?

Comment (janboll) [reply reconcile/utils/unleash.py:11]: I kept it the way it was. The previous implementation had the same behavior but destroyed the client after every request. 

Let me test and read up if the client is thread safe.

Comment (geoberle) [reply reconcile/utils/unleash.py]: thank you for taking additional time looking into this ❤️

File: reconcile/utils/unleash.py [hunk 3]
```diff
@@ -1,117 +1,96 @@
...
 
-from UnleashClient import UnleashClient
+from UnleashClient import UnleashClient, BaseCache
 from UnleashClient import strategies
+from UnleashClient.strategies import Strategy
 
-from reconcile.utils.defer import defer
 from reconcile.utils.helpers import toggle_logger
 
-log_lock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (janboll) [reply reconcile/utils/unleash.py:12]: I'll leave it removed...
https://docs.python.org/2/library/logging.html#thread-safety

**Review by fishi0x01 — APPROVED** [review]
Nice refactor! :+1:

**Pull request merged**